lib/husky_musher/utils/redcap.py

```diff
@@ -0,0 +1,529 @@
+    False
+
+    >>> need_to_create_new_td_for_today({'target': 1, 'complete_tos': None, 'complete_kr': 1})
+    True
+
+    >>> need_to_create_new_td_for_today({'target': 1, \
+        'complete_tos': get_todays_repeat_instance(), 'complete_kr': 1})
+    False
+
+    >>> need_to_create_new_td_for_today({'target': 1, \
+        'complete_tos': get_todays_repeat_instance(), 'complete_kr': None})
+    False
+    """
+    if not instances['target']:
+        return True
+
+    complete_tos_instance = instances['complete_tos']
+    if complete_tos_instance is None \
+        or (complete_tos_instance is not None \
+            and complete_tos_instance != get_todays_repeat_instance()):
+        if instances['complete_kr'] is not None:
+            return True
+
+    return False
+
+
+def need_to_create_new_kr_instance(instances: Dict[str, int]) -> bool:
+    """
+    Returns True if we need to create a new KR instance for the target TD
+    instance. Otherwise, returns False.
+
+    We need to create a new KR instance in the following conditions. Both of
+    these conditions assume a TD instance with [testing_trigger] = "Yes" exists
+    in the past 7 days.
+        1. No complete TOS instance exists on or after the target TD instance,
+           and no complete KR instance exists on or after the target TD instance.
+        2. A complete TOS instance exists on or after the target TD instance,
+           the TOS instance is not from today, and no complete KR instance
+           exists on or after the target TD instance.
+
+    *target_instance* is a TD instance number with [testing_trigger] = "Yes" in the
+    past 7 days.
+
+    *complete_tos_instance* is a TOS instance number marked complete on or after the
+    given *target_instance*.
+
+    *complete_kr_instance* is a KR instance number marked complete on or after the
+    given *target_intance*.
+
+    >>> need_to_create_new_kr_instance({'target': None, 'complete_tos': 1, 'complete_kr': 1})
+    False
+
+    >>> need_to_create_new_kr_instance({'target': 1, 'complete_tos': None, 'complete_kr': 1})
+    False
+
+    >>> need_to_create_new_kr_instance({'target': 1, 'complete_tos': 1, 'complete_kr': 1})
+    False
+
+    >>> need_to_create_new_kr_instance({'target': 1, 'complete_tos': None, 'complete_kr': None})
+    True
+
+    >>> need_to_create_new_kr_instance({'target': 1, 'complete_tos': None, 'complete_kr': 1})
+    False
+
+    >>> need_to_create_new_kr_instance({'target': 1, \
+        'complete_tos': get_todays_repeat_instance(), 'complete_kr': 1})
+    False
+
+    >>> need_to_create_new_kr_instance({'target': 1, \
+        'complete_tos': get_todays_repeat_instance(), 'complete_kr': None})
+    False
+    """
+    # Just to be safe, check to make sure we don't need to create a TD instance
+    # for today instead.
+    complete_tos_instance = instances['complete_tos']
+
+    if need_to_create_new_td_for_today(instances):
+        return False
```

Comment:
Should this raise an Exception?

Reply:
Ah, I see it would fall under the general logic error raised in `kiosk_registration_link`.